lib/net/tcp_client/tcp_client.rb

```diff
@@ -221,19 +227,27 @@ def self.connect(params={})
     #   puts "Received: #{response}"
     #   client.close
     def initialize(parameters={})
-      params                  = parameters.dup
-      @read_timeout           = (params.delete(:read_timeout) || 60.0).to_f
-      @connect_timeout        = (params.delete(:connect_timeout) || (@read_timeout/2)).to_f
-      buffered                = params.delete(:buffered)
-      @buffered               = buffered.nil? ? true : buffered
-      @connect_retry_count    = params.delete(:connect_retry_count) || 10
-      @retry_count            = params.delete(:retry_count) || 3
-      @connect_retry_interval = (params.delete(:connect_retry_interval) || 0.5).to_f
-      @on_connect             = params.delete(:on_connect)
-      @server_selector        = params.delete(:server_selector) || :ordered
-      @close_on_error         = params.delete(:close_on_error)
-      @close_on_error         = true if @close_on_error.nil?
-      @logger                 = params.delete(:logger)
+      params                      = parameters.dup
+      @read_timeout               = (params.delete(:read_timeout) || 60.0).to_f
+      @write_timeout              = (params.delete(:write_timeout) || -1).to_f
+      @connect_timeout            = (params.delete(:connect_timeout) || (@read_timeout/2)).to_f
+      buffered                    = params.delete(:buffered)
+      @buffered                   = buffered.nil? ? true : buffered
+      @connect_retry_count        = params.delete(:connect_retry_count) || 10
+      @retry_count                = params.delete(:retry_count) || 3
+      @connect_retry_interval     = (params.delete(:connect_retry_interval) || 0.5).to_f
+      @on_connect                 = params.delete(:on_connect)
+      @write_timeout              = (params[:write_timeout] || 60.0).to_f
+      @ssl                        = params.delete(:ssl)
+      @ssl                        = false if ssl.nil?
+      @ssl_connect_timeout        = (params.delete(:ssl_connect_timeout) || 30.0).to_f
+      @ssl_context_params         = (params.delete(:ssl_context_params) || {})
+      @ssl_connect_retry_count    = params.delete(:ssl_connect_retry_count) || 10
+      @ssl_connect_retry_interval = (params.delete(:ssl_connect_retry_interval) || 0.5).to_f
```

Comment:
Can we move the ssl options into a child hash within the options?
```ruby
Net::TCPClient.connect(
  servers:         ['server:3300', 'server2:3300'],
  server_selector: :random,
  ssl:             {
    cert:            cert,
    key:             key,
    key_pass_phrase: 'phrase',
    ca_cert:         ca_cert
  }
)
```

Rather than exposing the context directly, which requires a through understanding of the OpenSSL SSLContext, would the above options be sufficient?

I.e. Checkout https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/socket/ssl.rb

Does that cover enough options, or should we support both the above options and the context for further fine-tuning?